File: wasm-of-ocaml/benchmarks/C/funcrec.c
// https://github.com/Chris00/ocaml-benchmark/blob/master/examples/func_record.ml

typedef struct env Env;
struct env {
    int (*f)(int);
    int (*g)(int);
};

int h_struct(Env *e, int x){
    return 1 + e->f(x) + e->g(x);
}

int h_function(int (*f)(int), int (*g)(int), int x){
    return 1 + f(x) + g(x);
}

int f(int x){
    return x + 1;
}

int g(int x){
    return 2 * x;
}

int h(int x){
    return 1 + f(x) + g(x);
}

int iters = 10000;

int main(){
    int i;
    int a = 1, b = 1, c = 1;
    Env e = {&f, &g};
    // written as 3 loops to match ml version
    for (i = 0; i < iters; i++){
        a = h_struct(&e, a) % 1024;
    }
    for (i = 0; i < iters; i++){
        b = h_function(&f, &g, b) % 1024;
    }
    for (i = 0; i < iters; i++){
        c = h(c) % 1024;
    }
    return a+b+c;
}

File: wasm-of-ocaml/benchmarks/funcrec.ml
(* https://github.com/Chris00/ocaml-benchmark/blob/master/examples/func_record.ml
   Originally also had functor over a struct, and was based on floats rather than ints.
   Compares the performance of a simple function done different ways *)

type env = { f : int -> int;  g : int -> int }

let h_record e x = 1 + e.f x + e.g x

(* With arguments *)
let h_function f g x = 1 + f x + g x

let f x = x + 1
let g x = 2 * x

(* Without arguments *)
let h x = 1 + f x + g x

let record = {f = f; g = g} (* Avoid record being reallocated over and over *)

let rec iter f x = function
    | 0 -> x
    | n -> iter f ((f x) mod 1024) (n-1)

let _ = iter (h_record record) 1 10000
let _ = iter (h_function f g) 1 10000
let _ = iter h 1 10000

